stats/opencensus/stats.go

```diff
@@ -0,0 +1,227 @@
+		recordDataBegin(ctx, st)
+	case *stats.OutPayload:
+		recordDataOutPayload(ctx, st)
+	case *stats.InPayload:
+		recordDataInPayload(ctx, st)
+	case *stats.End:
+		recordDataEnd(ctx, st)
+	default:
+		// Shouldn't happen. gRPC calls into stats handler, and will never not
+		// be one of the types above.
+	}
+}
+
+// recordDataBegin takes a measurement related to the RPC beginning,
+// client/server started RPCs dependent on the caller.
+func recordDataBegin(ctx context.Context, b *stats.Begin) {
+	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
+	if !ok {
+		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
+		// context.
+		return
+	}
+
+	if b.IsClient() {
+		print("recording client started rpcs + 1")
+		ocstats.RecordWithOptions(ctx,
+			ocstats.WithTags(tag.Upsert(keyClientMethod, removeLeadingSlash(d.method))),
+			ocstats.WithMeasurements(clientStartedRPCs.M(1)))
+		return
+	}
+	ocstats.RecordWithOptions(ctx,
+		ocstats.WithTags(tag.Upsert(keyServerMethod, removeLeadingSlash(d.method))),
+		ocstats.WithMeasurements(serverStartedRPCs.M(1)))
+}
+
+// recordDataOutPayload records the length in bytes of outgoing messages and
+// increases total count of sent messages both stored in the RPCs (attempt on
+// client side) context for use in taking measurements at RPC end.
+func recordDataOutPayload(ctx context.Context, op *stats.OutPayload) {
+	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
+	if !ok {
+		// Shouldn't happen, as gRPC calls tagRPC which populates the rpcData in
+		// context.
+		return
+	}
+	atomic.AddInt64(&d.sentMsgs, 1)
+	atomic.AddInt64(&d.sentBytes, int64(op.Length))
+}
+
+// recordDataInPayload records the length in bytes of incoming messages and
+// increases total count of sent messages both stored in the RPCs (attempt on
+// client side) context for use in taking measurements at RPC end.
+func recordDataInPayload(ctx context.Context, ip *stats.InPayload) {
+	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
+	if !ok {
+		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
+		// context.
+		return
+	}
+	atomic.AddInt64(&d.recvMsgs, 1)
+	atomic.AddInt64(&d.recvBytes, int64(ip.Length))
+}
+
+// recordDataEnd takes per RPC measurements derived from information derived
+// from the lifetime of the RPC (RPC attempt client side).
+func recordDataEnd(ctx context.Context, e *stats.End) {
+	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
+	if !ok {
+		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
+		// context.
+		return
+	}
+	// latency bounds for distribution data (speced millisecond bounds) have
+	// fractions, thus need a float.
+	latency := float64(time.Since(d.startTime)) / float64(time.Millisecond)
+	var st string
+	if e.Error != nil {
+		s, _ := status.FromError(e.Error) // ignore second argument because codes.Unknown is fine
+		st = codes.CodeToStr[s.Code()]
+	} else {
+		st = "OK"
+	}
+
+	// TODO: Attach trace data through attachments?!?!
+
+	if e.Client {
```

Comment:
`stats.Begin` code calls `IsClient` but this checks the field (which is probably slightly better).  Unify them please.